### PR TITLE
add --lifecycleRule option with improved validation

### PR DESCRIPTION
**Pull request by mjurbanski-reef**
TODO
* [x] bump b2sdk requirement after https://github.com/reef-technologies/b2-sdk-python/pull/294  is merged and release - without it the validation of TypedDict will ALWAYS fail :<

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -109,6 +109,7 @@ def read_requirements(extra=None):
     # for example:
     # $ pip install -e .[dev,test]
     extras_require={
+        'full': read_requirements('full'),
```
Comment (kkalinowski-reef) [inline setup.py:112]: `full` means it's something missing from the original, while this is here only to provide a higher order validation of some inputs. How about `validation` or `typechecks`, or even just `checks`?

File: test/unit/test_console_tool.py [hunk 2]
```diff
@@ -188,7 +188,10 @@ def _run_command(
         expected_stderr = self._normalize_expected_output(expected_stderr, format_vars)
         stdout, stderr = self._get_stdouterr()
         console_tool = ConsoleTool(self.b2_api, stdout, stderr)
-        actual_status = console_tool.run_command(['b2'] + argv)
+        try:
+            actual_status = console_tool.run_command(['b2'] + argv)
+        except SystemExit as e:
```
Comment (kkalinowski-reef) [inline test/unit/test_console_tool.py:193]: Why is this needed? Doesn't normal `run_command` handle it properly?

File: b2/console_tool.py [hunk 3]
```diff
@@ -2537,9 +2566,8 @@ class UpdateBucket(DefaultSseMixin, Command):
 
     @classmethod
     def _setup_parser(cls, parser):
-        parser.add_argument('--bucketInfo', type=json.loads)
-        parser.add_argument('--corsRules', type=json.loads)
-        parser.add_argument('--lifecycleRules', type=json.loads)
+        parser.add_argument('--bucketInfo', type=validated_loads)
```
[the remark this answers is not in the document]
Comment (mjurbanski-reef) [reply b2/console_tool.py:2569]: What it does is it brings consistency. If we ever decide we want to swap how we `load` stuff we can do it in one place.

File: setup.py (hunk 1 above)
Comment (mjurbanski-reef) [reply setup.py:112]: oh, but it is missing, and we only allow installing without since we acknowledge some people may run into issues with the additional dependencies.

Then there is also this - Assume we add pycurl support, then if you are a user which want the best experience possible, and you don't care if there are binary dependencies or whatnot (and why would you as the user?), do you really want to do `pip install `b2[pycurl,validation]` ?

It seems like a common practice to group them (https://github.com/encode/uvicorn/blob/master/pyproject.toml#L36-L44 , https://github.com/pandas-dev/pandas/blob/main/pyproject.toml#L81 ) .
We could do add extra group for just validation boost, for these users that do really care about nitpicking their dependencies, but I'm not sure how big of group will that be.

File: test/unit/test_console_tool.py (hunk 2 above)
Comment (mjurbanski-reef) [reply test/unit/test_console_tool.py:193]: it does not 

our `run_command` typically just returns int, but the problem is that the `SystemExit` here comes from argparse - and I need to check that in my tests, hence this try except

File: setup.py (hunk 1 above)
Comment (kkalinowski-reef) [reply setup.py:112]: Oh, but in these cases all means that literally all dependencies are added, and each of them have a hefty list of optionals that can be added as well. So it should also include doc and license and curl and everything else in this full, shouldn't it? Currently it's just adding pydantic, which is used only for validation, isn't it?

Comment (kkalinowski-reef) [reply setup.py:112]: I'm not against `full`, or `all`. I'm just saying that this `full` is not really `full`.

File: test/unit/test_console_tool.py (hunk 2 above)
Comment (kkalinowski-reef) [reply test/unit/test_console_tool.py:193]: Ok, and from perspective of the user, it's just a `SystemExit`, so the `b2` just closes in the command line leaving a nice message behind and nothing "ugly", right? In that case, it's fine. I'm just surprised it's the first case in which we actually see `SystemExit` being raised.

Comment (mjurbanski-reef) [reply test/unit/test_console_tool.py:193]: `raise SystemExit(1)` does nothing else than just silenty ends the program and sets exit code to 1
the error message itself is printed out by argparse before that

Example:
```
$ b2 create-bucket --lifecycleRule '{}' --lifecycleRule '{"fileNamePrefix":"asd","daysFromUploadingToHiding":7,"daysFromHidingToDeleting":null}' test1231231231233 allPrivate
b2 create-bucket [-h] [--profile PROFILE] [--bucketInfo BUCKETINFO] [--corsRules CORSRULES] [--fileLockEnabled] [--replication REPLICATION]
                 [--defaultServerSideEncryption {SSE-B2,none}] [--defaultServerSideEncryptionAlgorithm {AES256}]
                 [--lifecycleRule LIFECYCLERULES | --lifecycleRules LIFECYCLERULES]
                 bucketName {allPublic,allPrivate}

  Creates a new bucket. Prints the ID of the bucket created.

  Optionally stores bucket info, CORS rules and lifecycle rules with the bucket. These can be given as JSON on the command line.

  If you want server-side encryption for all of the files that are uploaded to a bucket, you can enable SSE-B2 encryption as a default setting for the bucket. In order to
  do that pass --defaultServerSideEncryption=SSE-B2. The default algorithm is set to AES256 which can by changed with --defaultServerSideEncryptionAlgorithm parameter.
  All uploads to that bucket, from the time default encryption is enabled onward, will then be encrypted with SSE-B2 by default.

  To disable default bucket encryption, use --defaultServerSideEncryption=none.

  If --defaultServerSideEncryption is not provided, default server side encryption is determined by the server.

  Note

    Note that existing files in the bucket are not affected by default bucket encryption settings.

  Use --lifecycleRule to set lifecycle rule for the bucket. Multiple rules can be specified by repeating the option.

  --lifecycleRules option is deprecated and cannot be used together with --lifecycleRule.

  Requires capability:

    • writeBuckets
    • readBucketEncryption
    • writeBucketEncryption
    • writeBucketRetentions

positional arguments:
  bucketName
  {allPublic,allPrivate}

options:
  -h, --help            show this help message and exit
  --profile PROFILE
  --bucketInfo BUCKETINFO
  --corsRules CORSRULES
  --fileLockEnabled     If given, the bucket will have the file lock mechanism enabled. This parameter cannot be changed after bucket creation.
  --replication REPLICATION
  --defaultServerSideEncryption {SSE-B2,none}
  --defaultServerSideEncryptionAlgorithm {AES256}
  --lifecycleRule LIFECYCLERULES
                        Lifecycle rule in JSON format. Can be supplied multiple times.
  --lifecycleRules LIFECYCLERULES
                        (deprecated; use --lifecycleRule instead) List of lifecycle rules in JSON format.

b2 create-bucket: error: argument --lifecycleRule: Invalid value inputted, expected LifecycleRule (B2 Cloud Storage Lifecycle Rules <https://www.backblaze.com/docs/cloud-storage-lifecycle-rules>), got '{}', more detail below:
  In field 'fileNamePrefix' input was `{}`, error: Field required
```

**Comment by mjurbanski-reef** [comment]
merged upstream
